pkg/v1/layout/write.go

```diff
@@ -221,43 +223,83 @@ func (l Path) WriteFile(name string, data []byte, perm os.FileMode) error {
 // WriteBlob copies a file to the blobs/ directory in the Path from the given ReadCloser at
 // blobs/{hash.Algorithm}/{hash.Hex}.
 func (l Path) WriteBlob(hash v1.Hash, r io.ReadCloser) error {
+	return l.writeBlob(hash, 0, r, nil)
+}
+
+func (l Path) writeBlob(hash v1.Hash, size int64, r io.ReadCloser, renamer func() (v1.Hash, error)) error {
 	dir := l.path("blobs", hash.Algorithm)
 	if err := os.MkdirAll(dir, os.ModePerm); err != nil && !os.IsExist(err) {
 		return err
 	}
 
+	// Check if blob already exists and is the correct size
 	file := filepath.Join(dir, hash.Hex)
-	if _, err := os.Stat(file); err == nil {
-		// Blob already exists, that's fine.
+	if s, err := os.Stat(file); err == nil && (s.Size() == size || size == 0) {
 		return nil
 	}
+
+	// If a rename func was provided, use a temporary file suffix
```

Comment:
Is it possible to use io.TempFile for this?

Reply:
Sure, I have no preference.

The reason I had it using the digest (when available) plus a suffix was because it would be easy to watch a directory to see the progress of a large layer downloading - something I've commonly done.

With that in mind, would you like me to use the digest (when available) as part of the pattern - the second param to `ioutil.TempFile`?

Reply:
> The reason I had it using the digest (when available) plus a suffix was because it would be easy to watch a directory to see the progress of a large layer downloading - something I've commonly done.

I don't think I have a problem with pointing `TempFile` at the same directory (or a sibling/child directory) if that's useful to you.

> With that in mind, would you like me to use the digest (when available) as part of the pattern - the second param to ioutil.TempFile?

Yeah this is a great idea. We might find that doing some synchronization would be useful here as well (to avoid multiple writeLayer calls downloading the same blob)? We can just TODO that and follow up.

Reply:
I don't totally follow the synchronization comment. Are there any functions in the layout package that you think should be safe for concurrent use? Please suggest any TODOs and I'll add them.